pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java

```diff
@@ -0,0 +1,225 @@
+import org.apache.pinot.core.query.aggregation.groupby.ObjectGroupByResultHolder;
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.roaringbitmap.RoaringBitmap;
+
+
+public class ListAggFunction
+    extends BaseSingleInputAggregationFunction<ObjectArrayList<String>, String> {
+
+  private final String _separator;
+  private final boolean _nullHandlingEnabled;
+  private final boolean _isDistinct;
+
+  public ListAggFunction(ExpressionContext expression, String separator, boolean isDistinct,
+      boolean nullHandlingEnabled) {
+    super(expression);
+    _separator = separator;
+    _isDistinct = isDistinct;
+    _nullHandlingEnabled = nullHandlingEnabled;
+  }
+
+  @Override
+  public AggregationFunctionType getType() {
+    return AggregationFunctionType.LISTAGG;
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return new ObjectAggregationResultHolder();
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return new ObjectGroupByResultHolder(initialCapacity, maxCapacity);
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    ObjectArrayList<String> valueSet = aggregationResultHolder.getResult();
+    if (valueSet == null) {
+      valueSet = new ObjectArrayList<>();
+      aggregationResultHolder.setValue(valueSet);
+    }
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    String[] values = blockValSet.getStringValuesSV();
+    if (_nullHandlingEnabled) {
+      RoaringBitmap nullBitmap = blockValSet.getNullBitmap();
+      if (nullBitmap != null && !nullBitmap.isEmpty()) {
+        aggregateArrayWithNull(length, valueSet, values, nullBitmap);
+        return;
+      }
+    }
+    aggregateArray(length, valueSet, values);
+  }
+
+  private void aggregateArray(int length, ObjectArrayList<String> valueSet, String[] values) {
+    ObjectOpenHashSet<String> distinctValueSet = new ObjectOpenHashSet<>();
+    for (int i = 0; i < length; i++) {
+      String value = values[i];
+      if (_isDistinct) {
+        if (distinctValueSet.contains(value)) {
+          continue;
+        } else {
+          distinctValueSet.add(value);
+        }
+      }
+      valueSet.add(value);
+    }
+  }
+
+  private void aggregateArrayWithNull(int length, ObjectArrayList<String> valueSet, String[] values,
+      RoaringBitmap nullBitmap) {
+    ObjectOpenHashSet<String> distinctValueSet = new ObjectOpenHashSet<>();
+    for (int i = 0; i < length; i++) {
+      if (!nullBitmap.contains(i)) {
+        String value = values[i];
+        if (_isDistinct) {
+          if (distinctValueSet.contains(value)) {
+            continue;
+          } else {
+            distinctValueSet.add(value);
+          }
+        }
+        valueSet.add(value);
+      }
```

Comment:
Can you use the utility functions included in `NullableSingleInputAggregationFunction`? They have been proven to be faster than this direct approach when we check the bitmap on each iteration.

In this case you will probably need to use `foldNotNull`

Reply:
done, used `forEachNotNull`